<a href="https://colab.research.google.com/github/karthik-bha/IBM_data/blob/main/DV_with_Folium_and_Dashboard_with_Plotly_and_Dash.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install Folium

In [ ]:
import pandas as pd
import folium

In [ ]:
from folium.plugins import MarkerCluster
from folium.plugins import MousePosition
from folium.features import DivIcon

## Task 1: Mark all launch sites on a map

In [ ]:
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
spacex_df=pd.read_csv(url)
spacex_df.head()

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

In [ ]:
loc=spacex_df[ ['Lat','Long','Launch Site', 'class' ]]
ls_df=loc.groupby(['Launch Site'], as_index=False).first()
ls_df = ls_df[['Launch Site', 'Lat', 'Long']]
ls_df

Launch Site        Lat        Long
0   CCAFS LC-40  28.562302  -80.577356
1  CCAFS SLC-40  28.563197  -80.576820
2    KSC LC-39A  28.573255  -80.646895
3   VAFB SLC-4E  34.632834 -120.610745

In [ ]:
nasa_coord=[29.559684888503615, -95.0830971930759]
site_map=folium.Map(location=nasa_coord,zoom_start=10)

use folium.Circle to add a highlighted circle area with a text label on a specific coordinate.

In [ ]:
# Create a blue circle at NASA Johnson Space Center's coordinate with a popup label showing its name
# circle = folium.Circle(nasa_coord, radius=1000, color='#d35400', fill=True).add_child(folium.Popup('NASA Johnson Space Center'))
# # Create a blue circle at NASA Johnson Space Center's coordinate with a icon showing its name
# marker = folium.map.Marker( nasa_coord,
#     # Create an icon as a text label
# icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',))
# site_map.add_child(circle)
# site_map.add_child(marker)

In [ ]:
CCAFS_coord=[ls_df.loc[ 0,'Lat'], ls_df.loc[0, 'Long']]
CCAFS_LC=[ls_df.loc[ 1,'Lat'], ls_df.loc[1, 'Long']]
KSC_LC=[ls_df.loc[ 2,'Lat'], ls_df.loc[2, 'Long']]
VAFB_SLC=[ls_df.loc[ 3,'Lat'], ls_df.loc[3, 'Long']]
lis=[CCAFS_coord, CCAFS_LC, KSC_LC, VAFB_SLC]
lis2=['CCAFS_coord', 'CCAFS_LC', 'KSC_LC', 'VAFB_SLC']
colors=['blue','red','black','yellow']
y=0
c=0
for x in lis:
  print(lis2[y])
  # Create a blue circle
  circle = folium.Circle(x, radius=1000, color=c, fill=True)
  # Create a blue circle  with a icon showing its name
  marker = folium.map.Marker( x,
  # Create an icon as a text label
  icon=DivIcon(icon_size=(20,20),icon_anchor=(0,0),html='<div style="font-size: 12; color:c;"><b>%s</b></div>' % lis2[y]))
  site_map.add_child(circle)
  site_map.add_child(marker)
  y+=1
  c+=1

site_map


    Are all launch sites in proximity to the Equator line?
    Are all launch sites in very close proximity to the coast?
ans:
- Yes, most of them appear to be near the equator
- Yes, Most of them are near costal areas

# Task 2: Mark the success/failed launches for each site on the map


In [ ]:
spacex_df['marker_color'] = list(map(lambda x: 'green' if x==1 else 'red', spacex_df['class']))
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long marker_color  
46 -80.646895        green  
47 -80.646895        green  
48 -80.646895        green  
49 -80.576820        green  
50 -80.576820        green  
51 -80.576820          red  
52 -80.576820          red  
53 -80.576820          red  
54 -80.576820        green  
55 -80.576820          red

In [ ]:
def assign_marker_color(launch_outcome):
    if launch_outcome == 1:
        return 'green'
    else:
        return 'red'

spacex_df['marker_color'] = spacex_df['class'].apply(assign_marker_color)
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long marker_color  
46 -80.646895        green  
47 -80.646895        green  
48 -80.646895        green  
49 -80.576820        green  
50 -80.576820        green  
51 -80.576820          red  
52 -80.576820          red  
53 -80.576820          red  
54 -80.576820        green  
55 -80.576820          red

In [ ]:
marker_cluster = MarkerCluster()
# Add the Marker cluster to the site map
site_map.add_child(marker_cluster)
# for each row in spacex_df data frame
# create a Marker object with its coordinate
# and customize the Marker's icon property to indicate if this launch was successed or failed, e.g., icon=folium.Icon(color='white', icon_color=row['marker_color']
for site_lat, site_long, marker_color in zip(spacex_df['Lat'], spacex_df['Long'], spacex_df['marker_color']):
    site_coordinate = [site_lat, site_long]
    marker = folium.map.Marker(
        site_coordinate,
        # Create an icon as a text label
        icon=folium.Icon(color='white',
                         icon_color=marker_color)
    )
    marker.add_to(marker_cluster)

site_map



# TASK 3: Calculate the distances between a launch site to its proximities


In [ ]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

In [ ]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

In [ ]:


# distance_railway = calculate_distance(lat1, lon1, lat2, lon2)
railway_marker = [28.55752, -80.80155]
launch_coordinate = [28.57337, -80.64669]
distance_railway = calculate_distance(railway_marker[0], railway_marker[1], launch_coordinate[0], launch_coordinate[1])
distance_railway  # distance in km



15.230651245141603

In [ ]:


# create and add a folium.Marker on your selected closest raiwaly point on the map
# show the distance to the launch site using the icon property
marker = folium.map.Marker(
        railway_marker,
        # Create an icon as a text label
        icon=DivIcon(
            icon_size=(400, 400),
            icon_anchor=(0, 0),
            html='<div style="font-size:400; color:#0c10f2;"><b>%s</b></div>' % str(round(distance_railway, 2))+' km',
            )
    )
marker.add_to(site_map)

site_map



In [ ]:
# Create a `folium.PolyLine` object using the railway point coordinate and launch site coordinate
folium.PolyLine([railway_marker, launch_coordinate], color='blue').add_to(site_map)
site_map


In [ ]:


# Create a marker with distance to a closest city, coastline, highway, etc.
# Draw a line between the marker to the launch site
city   = [28.61200, -80.80788]
coastline = [28.5858, -80.79952]
highway   = [28.5402, -80.85079]

city_distance = calculate_distance(city[0], city[1], launch_coordinate[0], launch_coordinate[1])
coastline_distance = calculate_distance(coastline[0], coastline[1], launch_coordinate[0], launch_coordinate[1])
highway_distance = calculate_distance(highway[0], highway[1], launch_coordinate[0], launch_coordinate[1])

colors = ['red','orange','green']
html_colors = ['#dc3545','#fd7e14','#198754']

for coordinate ,distance, color, html_color in zip([city, coastline, highway], [city_distance, coastline_distance, highway_distance], colors, html_colors):
    marker = folium.map.Marker(
            coordinate,
            # Create an icon as a text label
            icon=DivIcon(
                icon_size=(20,20),
                icon_anchor=(0,0),
                html='<div style="font-size: 12; color:'+html_color+';"><b>%s</b></div>' % str(round(distance, 2)) + 'km',
                )
            )
    marker.add_to(site_map)
    folium.PolyLine([coordinate, launch_coordinate], color=color).add_to(site_map)
site_map




    Are launch sites in close proximity to railways?
    Are launch sites in close proximity to highways?
    Are launch sites in close proximity to coastline?
    Do launch sites keep certain distance away from cities?



    - From the visual analysis of the launch site KSC LC-39A we can clearly see that it is:
        -relatively close to railway (15.23 km)
        -relatively close to highway (20.28 km)
        -relatively close to coastline (14.99 km)
   - Also the launch site KSC LC-39A is relative close to its closest city Titusville (16.32 km).
    - Failed rocket with its high speed can cover distances like 15-20 km in few seconds. It could be potentially dangerous to populated areas.


# DV with plotly

In [ ]:
# Import required libraries
import pandas as pd
#!pip install dash
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output
import plotly.express as px

# Read the airline data into pandas dataframe
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'
spacex_df=pd.read_csv(url)
max_payload = spacex_df['Payload Mass (kg)'].max()
min_payload = spacex_df['Payload Mass (kg)'].min()

# Create a dash application
app = dash.Dash(__name__)

# Create an app layout
app.layout = html.Div(children=[html.H1('SpaceX Launch Records Dashboard',
                                        style={'textAlign': 'center', 'color': '#503D36',
                                               'font-size': 40}),
                                # TASK 1: Add a dropdown list to enable Launch Site selection
                                # The default select value is for ALL sites
                                dcc.Dropdown(id='site-dropdown',
                                options=[
                                    {'label': 'All Sites', 'value': 'All Sites'},
                                    {'label': 'CCAFS LC-40', 'value': 'CCAFS LC-40'},
                                    {'label': 'VAFB SLC-4E', 'value': 'VAFB SLC-4E'},
                                    {'label': 'KSC LC-39A', 'value': 'KSC LC-39A'},
                                    {'label': 'CCAFS SLC-40', 'value': 'CCAFS SLC-40'}
                                ],
                                placeholder='Select a Launch Site Here',
                                value='All Sites',
                                searchable=True
                                ),
                                html.Br(),

                                # TASK 2: Add a pie chart to show the total successful launches count for all sites
                                # If a specific launch site was selected, show the Success vs. Failed counts for the site
                                html.Div(dcc.Graph(id='success-pie-chart')),
                                html.Br(),

                                html.P("Payload range (Kg):"),
                                # TASK 3: Add a slider to select payload range
                                dcc.RangeSlider(id='payload-slider',
                                min=0,
                                max=10000,
                                step=1000,
                                marks={i: '{}'.format(i) for i in range(0, 10001, 1000)},
                                value=[min_payload, max_payload]),

                                # TASK 4: Add a scatter chart to show the correlation between payload and launch success
                                html.Div(dcc.Graph(id='success-payload-scatter-chart')),
                                ])

# TASK 2:
# Add a callback function for `site-dropdown` as input, `success-pie-chart` as output
@app.callback( Output(component_id='success-pie-chart', component_property='figure'),
               Input(component_id='site-dropdown', component_property='value'))
def get_pie_chart(launch_site):
    if launch_site == 'All Sites':
        fig = px.pie(values=spacex_df.groupby('Launch Site')['class'].mean(),
                     names=spacex_df.groupby('Launch Site')['Launch Site'].first(),
                     title='Total Success Launches by Site')
    else:
        fig = px.pie(values=spacex_df[spacex_df['Launch Site']==str(launch_site)]['class'].value_counts(normalize=True),
                     names=spacex_df['class'].unique(),
                     title='Total Success Launches for Site {}'.format(launch_site))
    return(fig)

# TASK 4:
# Add a callback function for `site-dropdown` and `payload-slider` as inputs, `success-payload-scatter-chart` as output
@app.callback( Output(component_id='success-payload-scatter-chart', component_property='figure'),
              [Input(component_id='site-dropdown', component_property='value'),
               Input(component_id='payload-slider',component_property='value')])
def get_payload_chart(launch_site, payload_mass):
    if launch_site == 'All Sites':
        fig = px.scatter(spacex_df[spacex_df['Payload Mass (kg)'].between(payload_mass[0], payload_mass[1])],
                x="Payload Mass (kg)",
                y="class",
                color="Booster Version Category",
                hover_data=['Launch Site'],
                title='Correlation Between Payload and Success for All Sites')
    else:
        df = spacex_df[spacex_df['Launch Site']==str(launch_site)]
        fig = px.scatter(df[df['Payload Mass (kg)'].between(payload_mass[0], payload_mass[1])],
                x="Payload Mass (kg)",
                y="class",
                color="Booster Version Category",
                hover_data=['Launch Site'],
                title='Correlation Between Payload and Success for Site {}'.format(launch_site))
    return(fig)


# Run the app
if __name__ == '__main__':
    app.run_server()

<IPython.core.display.Javascript object>